tests/filter_kalman.cpp

```diff
@@ -0,0 +1,202 @@
+#include ".test.h"
+
+#include "xmath.h"
+#include <random>
+#include <fstream>
+#include <sstream>
+
+#include "ascii.h"
+
+using namespace ascii;
+using namespace xmath;
+
+
+constexpr auto dt = 0.1;
+
+mat<2, 2> stm_cart = {
+	{ 1, dt },
+	{ 0, 1  }
+};
+
+mat<2, 2> process_noise_covar_cart = {
+	{ 0.01, 0 },
+	{ 0, 0.01 },
+};
+
+mat<2, 1> control_to_state_cart = { { 0 }, {dt} };
+
+mat<1, 2> state_to_measurement_cart = {
+	{ 1, 0 },
+};
+
+mat<1, 1> measurement_noise_covar_cart = { 
+	{ 0.1 }
+};
+
+std::default_random_engine rng;
+
+
+void random_walk_velocity()
+{
+
+	std::normal_distribution<double> process_noise(0,0.01);
+	std::normal_distribution<double> measurement_noise(0.0,0.1);
+	std::normal_distribution<double> wander_noise(-1,1);
+
+	vec<2> state = {}; // consists of 1D position, and velocity
+
+	filter::kalman<1, 2, 1> cart_filter(state); 
+	std::unordered_map<std::string, std::vector<double>> traces;
+
+	int i = 0;
+
+	state[1] = 1;
+
+	for (float t = 0; t < 10; t += dt)
+	{
+		state[1] += wander_noise(rng);
+		state = stm_cart * state + vec<2>{process_noise(rng),process_noise(rng)};
+
+		if (i % 10 == 0)
+		{
+			auto z = state[0] + measurement_noise(rng);
+
+
+			cart_filter.measurement_update(
+				stm_cart,
+				state_to_measurement_cart,
+				mat<1, 1>{{z}},
+				measurement_noise_covar_cart
+			);
+		}
+
+		cart_filter.time_update(stm_cart, control_to_state_cart, {}, process_noise_covar_cart);
+
+		traces["true pos"].push_back(state[0]);
+		traces["true vel"].push_back(state[1]);
+		traces["estimated pos"].push_back(cart_filter.estimated.state[0][0]);
+		traces["estimated vel"].push_back(cart_filter.estimated.state[1][0]);
+		i += 1;
+	}
+
+	Asciichart chart(traces);
+	std::cerr << __FUNCTION__ << std::endl;
+	std::cerr << chart.offset(3).height(30).legendPadding(3).Plot();
+
+	// ensure each state element is within 1-sigma of the true state
+	std::cerr << cart_filter.estimated.covariance.to_string() << std::endl;
+	const auto& cov = cart_filter.estimated.covariance;
+	assert(fabs(cart_filter.estimated.state[0][0] - state[0]) < sqrt(cov[0][0]));
```

Comment:
This test fails for me:

```
filter_kalman: filter_kalman.cpp:89: void random_walk_velocity(): Assertion `fabs(cart_filter.estimated.state[0][0] - state[0]) < sqrt(cov[0][0])' failed.
./bin/filter_kalman...FAIL
```

Here's the plot I get:

```
random_walk_velocity
estimated vel  0 ┼───────────╮─╮─────╮                                                                                
estimated pos-18 ┤      ╰────╰───────╮╮─╭─╮──────────────╮                                                            
true vel     -36 ┤                   ╰──────╮──────────╮ ╰───────────────────╮                                        
true pos     -54 ┤                          ╰────╮╮    ╰───────────────╮     ╰─────────╮                              
             -73 ┤                               ╰───╮───╮             ╰─────────────────────────╮─────────────────── 
             -91 ┤                                   ╰───╮                                       ╰─────────────────── 
             -1.1e+02 ┤                                       ╰───╮─╮                                                      
             -1.3e+02 ┤                                           ╰──╮──╮                                                  
             -1.4e+02 ┤                                              ╰──╮                                                  
             -1.6e+02 ┤                                                 ╰──╮╮                                              
             -1.8e+02 ┤                                                    ╰──╮─╮                                          
             -2e+02 ┤                                                       ╰──╮╮                                        
             -2.2e+02 ┤                                                          ╰─╮                                       
             -2.4e+02 ┤                                                            ╰──╮╮                                   
             -2.5e+02 ┤                                                               ╰─╮─╮                                
             -2.7e+02 ┤                                                                 ╰──╮╮                              
             -2.9e+02 ┤                                                                    ╰─╮                             
             -3.1e+02 ┤                                                                      ╰──╮                          
             -3.3e+02 ┤                                                                         ╰─╮╮                       
             -3.4e+02 ┤                                                                           ╰─╮╮                     
             -3.6e+02 ┤                                                                             ╰─╮                    
             -3.8e+02 ┤                                                                               ╰─╮                  
             -4e+02 ┤                                                                                 ╰──╮               
             -4.2e+02 ┤                                                                                    ╰─╮             
             -4.4e+02 ┤                                                                                      ╰─╮╮          
             -4.5e+02 ┤                                                                                        ╰─╮         
             -4.7e+02 ┤                                                                                         ╰╰─╮       
             -4.9e+02 ┤                                                                                            ╰─╮     
             -5.1e+02 ┤                                                                                              ╰─╮   
             -5.3e+02 ┤                                                                                                ╰─╮ 
             -5.4e+02 ┤                                                                                                  ╰ 
|0.440498, 0.232486|
|0.232486, 0.244473|

```

Hmm, the plot isn't clear at all in black and white. Basically, the position estimate tracks the real position quite well, but the velocity seems to be delayed a tiny bit.

Reply:
Huh, not sure why it looks like that. This is what you should be seeing. Yeah, I'm not making any claims that any of these filters are well tuned 😆 
![Screenshot from 2022-09-17 16-43-30](https://user-images.githubusercontent.com/1676878/190878711-a707d7a7-2c75-4ddb-aa32-374ee67e1706.png)